### DISCUSSION TITLE
Comparisons between CUE, Jsonnet, Dhall, OPA, etc.

### DISCUSSION THREAD

**Discussion by cueckoo** [cue-lang/cue#669] [Q&A] [answered]
_Originally opened by @klauern in https://github.com/cuelang/cue/discussions/669_

Are there any references to help explain the differences between CUE and some tools that I believe (feel free to correct me if I'm off-base) are similar?

* JSONNET
* Dhall
* OPA (as well as Conftest)

I'm completely unfamiliar with CUE, but I have seen the above three used for generating YAML for Kubernetes, or evaluate compliance of those against internal policies.  Is CUE a good fit for that kind of work?  Is it suited for other kinds of work?  I see a number of items in the [Use Cases](https://cuelang.org/docs/usecases/) that make me think that CUE is in the same space, but I am not sure if it's worth looking into it at this time considering I have some experience with the above three.  Is there a page or links to sites that help explain where CUE shines or what would make it stand out compared to those?

**Comment by cueckoo** [comment, marked as answer]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/669#discussioncomment-306811_

CUE was developed as an answer to the pitfalls of GCL (very much like Jsonnet).  See the discussion in https://cuelang.org/docs/usecases/configuration/. It is based on an 30-year old approach from NLP, where it was used to model large 100k lines up configurations (grammars and lexicons), maintained by distributed groups of teams.

Dhall addresses some of the issues of GCL and Jsonnet (like lack of typing), but lacks the detailed typing of CUE. But it still misses the most important property of CUE: its model of composability. Some of the benefits are explained in the above link.  Conceptually, CUE is an aspect-oriented and constraint-based language. It allows you to specify fine-grained constraints on what are valid values. These constraints then double as templates, allowing to remove boilerplate often with the same efficacy as inheritance, even if it works very differently.  

The lack of override inheritance in CUE also makes it much easier to see where values are coming from. A value is always final in CUE, it can only be made more specific. This is in stark contrast with inheritance or other forms of composition, where values can often be overridden many times, making it very hard to see where values come from. CUE's approach leads to big improvements in maintainability. It may be harder to write initially, partly because its conceptually so different from the other languages, but it is considerably easier to maintain. 

I would argue that for configuration languages maintainability and readability are more important even than for programming languages, because they are ofter viewed by a larger group, often need to be changed in emergency conditions, and also as they are supposed to convey a certain contract. Most configuration languages, like GCL (my own doing), are more like scripting languages, making it easier to crank out definitions of large swats of data compactly, but being harder to comprehend and modify later.

CUE also can be used for policy specification, like Rego (OPA).CUE unifies values, types, and constraints in a single continuum. As it is  a constraint-based language first and foremost, it is well suited for defining policy. It is less developed in that area than Rego, but it I expect it will ultimately be better suited for policy. Note that Rego is based on Datalog, which is more of a query language at hart, giving it quite a different feel for defining policy than CUE. Both are logic programming languages, though, and share many of the same properties.

OPA itself, though, is more of a product. CUE is more like Rego, the language used in OPA. It could make sense, for instance, to have a CUE frontend to OPA. 

I hope this helps a bit.

**Reply by samlaf** [reply]
@mpvl thank you for this detailed writeup. I keep coming back to it and learning new things everytime, and wish you had written more! I want to expend (and get your review) my current understanding of what you mean by "CUE's model for composability", and how it compares to inheritance's model for "composability".

Whenever I hear "algebraic model for composability", I immediately think about monads. So I was surprised when reading that CUE's core engine also uses an algebraic structure for composability: a (semi)lattice. My current take/understanding is that:
- monads (monoids aka group-like) structures are for CONTROL-FLOW composition
- lattices are for DATA composition

Or similarly:
- Monads: "Here's HOW to compute it" (imperative)
- Lattices: "Here's WHAT it should satisfy" (declarative)

The main idea being that monads help you SYNTACTICALLY deal with composition of functions that have (side) effects, but not SEMANTICALLY (can still shoot yourself in the foot because function composition is not commutative or idempotent). On the other hand, lattices are much more constraining, and for that, gain commutativity and idempotency. This is why they are used in things like config languages (CUE) and data pipelines (CRDTs)!

Inheritance, in the context of pure data inheritance (not OOP inheritance with execution model), is not a lattice, but is instead a monoid:
<img width="645" alt="image" src="https://github.com/user-attachments/assets/7559750e-297f-436f-9a3b-29705c47b04d" />

Ted Kaminski has already written about [multiple forms of composition](https://www.tedinski.com/2018/05/15/familiar-forms-of-composition.html), but there's no mention of CUE or "mixed value and type hierarchies" in there (other than dependent types.. which might be related?). I keep trying to build an analogy between what CUE is doing to configs, and what type/trait inference/constraint systems (like Rust's [chalk](https://rustc-dev-guide.rust-lang.org/traits/chalk.html)) are doing to function/library composition, because they feel very similar.

Microservice architectures threw out type-safe library composition by putting a serialization and language barrier between their services. But it feels like CUE can give similar properties to binary orchestrators (say devops engineers who manage a k8s cluster) by letting them compose different binaries (similar to how modular monolith developers can compose different libraries safely using type systems that unify their generic functions) in a safe way:
```
// "Libraries" are constraint templates
#Database: {
    host: string
    port: int & >1000
    ssl: bool
}

#WebServer: {
    port: int & >8000  
    timeout: int & >0
}

#SecurityPolicy: {
    ssl: true
    port: int & >8000  // Only secure ports
}

// Composition guided by constraint compatibility
myApp: #Database & #WebServer & #SecurityPolicy & {
    host: "localhost"
    port: 8443        // Must satisfy ALL port constraints
    timeout: 30
    ssl: true
}
```
Just like how Rust's type system prevents you from passing a String where an Int is expected, CUE prevents you from composing incompatible configurations.
